sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeOuterJoin.scala

```diff
@@ -0,0 +1,249 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.joins
+
+import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans.{JoinType, LeftOuter, RightOuter}
+import org.apache.spark.sql.catalyst.plans.physical._
+import org.apache.spark.sql.execution.{BinaryNode, SparkPlan}
+
+/**
+ * :: DeveloperApi ::
+ * Performs an sort merge outer join of two child relations.
+ *
+ * Note: this does not support full outer join yet; see SPARK-9730 for progress on this.
+ */
+@DeveloperApi
+case class SortMergeOuterJoin(
+    leftKeys: Seq[Expression],
+    rightKeys: Seq[Expression],
+    joinType: JoinType,
+    condition: Option[Expression],
+    left: SparkPlan,
+    right: SparkPlan) extends BinaryNode {
+
+  override def output: Seq[Attribute] = {
+    joinType match {
+      case LeftOuter =>
+        left.output ++ right.output.map(_.withNullability(true))
+      case RightOuter =>
+        left.output.map(_.withNullability(true)) ++ right.output
+      case x =>
+        throw new IllegalArgumentException(
+          s"${getClass.getSimpleName} should not take $x as the JoinType")
+    }
+  }
+
+  override def outputPartitioning: Partitioning = joinType match {
+    case LeftOuter => left.outputPartitioning
+    case RightOuter => right.outputPartitioning
+    case x =>
+      throw new IllegalArgumentException(
+        s"${getClass.getSimpleName} should not take $x as the JoinType")
+  }
+
+  override def outputOrdering: Seq[SortOrder] = joinType match {
+    case LeftOuter => requiredOrders(leftKeys)
+    case RightOuter => requiredOrders(rightKeys)
+    case x => throw new IllegalArgumentException(
+      s"SortMergeOuterJoin should not take $x as the JoinType")
+  }
+
+  override def requiredChildDistribution: Seq[Distribution] =
+    ClusteredDistribution(leftKeys) :: ClusteredDistribution(rightKeys) :: Nil
+
+  override def requiredChildOrdering: Seq[Seq[SortOrder]] =
+    requiredOrders(leftKeys) :: requiredOrders(rightKeys) :: Nil
```

Comment:
I don't think we need to worry about sort regression for joins right now at
all. Just switch to sorting for shuffle join.

For aggregation the problem is that aggregation leads to data reduction,
and hash based aggregation exploits that fact, whereas sort based
aggregation doesn't and that's why it is really important to start with
hash and fallback to sort.

In join however it is very unclear whether sort vs hash is better. There
have been a lot of papers coming out about this topic and it has never been
conclusive.

On Saturday, August 8, 2015, Davies Liu <notifications@github.com> wrote:

> In
> sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeOuterJoin.scala
> <https://github.com/apache/spark/pull/7904#discussion_r36581693>:
>
> > +      throw new IllegalArgumentException(
> > +        s"${getClass.getSimpleName} should not take $x as the JoinType")
> > +  }
> > +
> > +  override def outputOrdering: Seq[SortOrder] = joinType match {
> > +    case LeftOuter => requiredOrders(leftKeys)
> > +    case RightOuter => requiredOrders(rightKeys)
> > +    case x => throw new IllegalArgumentException(
> > +      s"SortMergeOuterJoin should not take $x as the JoinType")
> > +  }
> > +
> > +  override def requiredChildDistribution: Seq[Distribution] =
> > +    ClusteredDistribution(leftKeys) :: ClusteredDistribution(rightKeys) :: Nil
> > +
> > +  override def requiredChildOrdering: Seq[Seq[SortOrder]] =
> > +    requiredOrders(leftKeys) :: requiredOrders(rightKeys) :: Nil
>
> Yes, it's safe to choose sort. For aggregation, we have to introduce
> hybrid aggregation to address the performance regression from sort based
> aggregation. We may also should do this for join in next release.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/7904/files#r36581693>.
>